### ISSUE TITLE
Test failures with python 3.10

### ISSUE THREAD

**Issue by msarahan** [guildai/guildai#330]
Vendored pkg_resources thinks python 3.10 < 3.5:
```
**********************************************************************
File "/home/msarahan/code/guildai/guild/tests/deprecated-options.md", line 30, in deprecated-options.md
Failed example:
    test("guild ls --unlabeled")
Expected:
    WARNING: option --unlabeled is deprecated and will be removed in version 0.8 - use '--label -' instead
    guild: no matching runs
    <exit 1>
Got:
    WARNING: option --unlabeled is deprecated and will be removed in version 0.8 - use '--label -' instead
    Traceback (most recent call last):
      File "/home/msarahan/code/guildai/guild/external/pkg_resources/__init__.py", line 584, in _build_master
        ws.require(__requires__)
      File "/home/msarahan/code/guildai/guild/external/pkg_resources/__init__.py", line 901, in require
        needed = self.resolve(parse_requirements(requirements))
      File "/home/msarahan/code/guildai/guild/external/pkg_resources/__init__.py", line 792, in resolve
        raise VersionConflict(dist, req).with_context(dependent_req)
    pkg_resources.ContextualVersionConflict: (psutil 5.9.0 (/home/msarahan/mambaforge/envs/guildai/lib/python3.10/site-packages), Requirement.parse('psutil<5.8,>=5.6.3'), {'guildai'})
    <BLANKLINE>
    During handling of the above exception, another exception occurred:
    <BLANKLINE>
    Traceback (most recent call last):
      File "/home/msarahan/mambaforge/envs/guildai/bin/guild", line 33, in <module>
        sys.exit(load_entry_point('guildai', 'console_scripts', 'guild')())
      File "/home/msarahan/code/guildai/guild/main_bootstrap.py", line 40, in main
        _main()
      File "/home/msarahan/code/guildai/guild/main_bootstrap.py", line 66, in _main
        guild.main.main()
      File "/home/msarahan/code/guildai/guild/main.py", line 33, in main
        main_cmd.main(standalone_mode=False)
      File "/home/msarahan/mambaforge/envs/guildai/lib/python3.10/site-packages/click/core.py", line 829, in __call__
        return self.main(*args, **kwargs)
      File "/home/msarahan/mambaforge/envs/guildai/lib/python3.10/site-packages/click/core.py", line 782, in main
        rv = self.invoke(ctx)
      File "/home/msarahan/mambaforge/envs/guildai/lib/python3.10/site-packages/click/core.py", line 1259, in invoke
        return _process_result(sub_ctx.command.invoke(sub_ctx))
      File "/home/msarahan/mambaforge/envs/guildai/lib/python3.10/site-packages/click/core.py", line 1066, in invoke
        return ctx.invoke(self.callback, **ctx.params)
      File "/home/msarahan/mambaforge/envs/guildai/lib/python3.10/site-packages/click/core.py", line 610, in invoke
        return callback(*args, **kwargs)
      File "/home/msarahan/mambaforge/envs/guildai/lib/python3.10/site-packages/click/decorators.py", line 21, in new_func
        return f(get_current_context(), *args, **kwargs)
      File "/home/msarahan/code/guildai/guild/click_util.py", line 213, in fn
        return fn0(*(args + (Args(**kw),)))
      File "/home/msarahan/code/guildai/guild/commands/ls.py", line 96, in ls
        from . import ls_impl
      File "/home/msarahan/code/guildai/guild/commands/ls_impl.py", line 26, in <module>
        from . import remote_impl_support
      File "/home/msarahan/code/guildai/guild/commands/remote_impl_support.py", line 21, in <module>
        from guild import remote as remotelib
      File "/home/msarahan/code/guildai/guild/remote.py", line 21, in <module>
        from guild import entry_point_util
      File "/home/msarahan/code/guildai/guild/entry_point_util.py", line 17, in <module>
        import pkg_resources
      File "/home/msarahan/code/guildai/guild/external/pkg_resources/__init__.py", line 3254, in <module>
        def _initialize_master_working_set():
      File "/home/msarahan/code/guildai/guild/external/pkg_resources/__init__.py", line 3237, in _call_aside
        f(*args, **kwargs)
      File "/home/msarahan/code/guildai/guild/external/pkg_resources/__init__.py", line 3266, in _initialize_master_working_set
        working_set = WorkingSet._build_master()
      File "/home/msarahan/code/guildai/guild/external/pkg_resources/__init__.py", line 586, in _build_master
        return cls._build_from_requirements(__requires__)
      File "/home/msarahan/code/guildai/guild/external/pkg_resources/__init__.py", line 599, in _build_from_requirements
        dists = ws.resolve(reqs, Environment())
      File "/home/msarahan/code/guildai/guild/external/pkg_resources/__init__.py", line 787, in resolve
        raise DistributionNotFound(req, requirers)
    pkg_resources.DistributionNotFound: The 'scikit-learn<0.21,>=0.19.1; python_version < "3.5"' distribution was not found and is required by guildai
    <exit 1>
**********************************************************************
```

We probably just need to port whatever fix pkg_resources currently has.

**Comment by msarahan** [comment]
Moving along with trying to get python 3.10 working. When I replace the vendored pkg_resources and setuptools (which are merged now), things mostly work but a couple more errors pop up:

Setuptools appears to override distutils, but is not able to do so in our test suite:
```
**********************************************************************
File "/home/msarahan/code/guildai/guild/tests/imports.md", line 25, in imports.md
Failed example:
    for name in sorted([m.__name__ for m in iter_mods()]):
      print(name) # doctest: +REPORT_UDIFF
Exception raised:
    Traceback (most recent call last):
      File "/home/msarahan/mambaforge/envs/guildai/lib/python3.10/doctest.py", line 1346, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest imports.md[6]>", line 1, in <module>
        for name in sorted([m.__name__ for m in iter_mods()]):
      File "<doctest imports.md[6]>", line 1, in <listcomp>
        for name in sorted([m.__name__ for m in iter_mods()]):
      File "<doctest imports.md[5]>", line 14, in iter_mods
        yield importlib.import_module(mod_name)
      File "/home/msarahan/mambaforge/envs/guildai/lib/python3.10/importlib/__init__.py", line 126, in import_module
        return _bootstrap._gcd_import(name[level:], package, level)
      File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
      File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
      File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
      File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
      File "<frozen importlib._bootstrap_external>", line 883, in exec_module
      File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
      File "/home/msarahan/code/guildai/guild/package_main.py", line 27, in <module>
        import setuptools
      File "/home/msarahan/mambaforge/envs/guildai/lib/python3.10/site-packages/setuptools/__init__.py", line 8, in <module>
        import _distutils_hack.override  # noqa: F401
      File "/home/msarahan/mambaforge/envs/guildai/lib/python3.10/site-packages/_distutils_hack/override.py", line 1, in <module>
        __import__('_distutils_hack').do_override()
      File "/home/msarahan/mambaforge/envs/guildai/lib/python3.10/site-packages/_distutils_hack/__init__.py", line 72, in do_override
        ensure_local_distutils()
      File "/home/msarahan/mambaforge/envs/guildai/lib/python3.10/site-packages/_distutils_hack/__init__.py", line 59, in ensure_local_distutils
        assert '_distutils' in core.__file__, core.__file__
    AssertionError: /home/msarahan/mambaforge/envs/guildai/lib/python3.10/distutils/core.py
**********************************************************************
```

Error text has changed (just need to update tests to account for this:
```
**********************************************************************
File "/home/msarahan/code/guildai/guild/tests/notebooks.md", line 259, in notebooks.md
Failed example:
    logs.print_all()
Expected:
    WARNING: error parsing Python source "'unterminated string": EOL
    while scanning string literal (<unknown>, line 1)
Got:
    WARNING: error parsing Python source "'unterminated string": unterminated string literal (detected at line 1) (<unknown>, line 1)
**********************************************************************
```

Method replacement in python_utils doesn't appear to be working:
```
**********************************************************************
File "/home/msarahan/code/guildai/guild/tests/python-utils.md", line 335, in python-utils.md
Failed example:
    calc.incr(1, 2)
Expected:
    Traceback (most recent call last):
    TypeError: incr() takes ...
Got:
    Traceback (most recent call last):
      File "/home/msarahan/mambaforge/envs/guildai/lib/python3.10/doctest.py", line 1346, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest python-utils.md[58]>", line 1, in <module>
        calc.incr(1, 2)
    TypeError: Calc.incr() takes 2 positional arguments but 3 were given
**********************************************************************
```

I believe this one corresponds to this example:
```
Here's a function that supports a second argument, which specifies the
amount to increment by. Note we return a result by raising
`python_util.Result` (as in the previous example). Note we don't use
the original method (represented by the `_incr` argument) because
we're replacing it altogether.

    >>> def incr2(_incr, x, incr_by=1):
    ...   raise python_util.Result(x + incr_by)

We wrap the original:

    >>> python_util.listen_method(Calc, "incr", incr2)
    <guild.python_util.MethodWrapper object at ...>

And here's our new behavior:

    >>> calc.incr(1)
    2
    >>> calc.incr(1, 2)
    3
```

Finally, VCS stuff got messed up somehow. This might have been a timeout issue with the tests waiting on me to type in my password, as I had started the tests and walked away for lunch.

```
**********************************************************************
File "/home/msarahan/code/guildai/guild/tests/vcs-utils.md", line 56, in vcs-utils.md
Failed example:
    commit, status = vcs_util.commit_for_dir(repo)
Exception raised:
    Traceback (most recent call last):
      File "/home/msarahan/mambaforge/envs/guildai/lib/python3.10/doctest.py", line 1346, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest vcs-utils.md[13]>", line 1, in <module>
        commit, status = vcs_util.commit_for_dir(repo)
      File "/home/msarahan/code/guildai/guild/vcs_util.py", line 89, in commit_for_dir
        raise NoCommit(dir)
    guild.vcs_util.NoCommit: /tmp/guild-test-_0uzp3wr
**********************************************************************
```

**Comment by msarahan** [comment]
Talked with Garrett, and we'll drop python <=3.5 so we can remove the requirements.txt lines that are causing issues. The other issues require more investigation.

**Comment by daoudclarke** [comment]
I was really keen to try out Guild AI but my project is using 3.10. It's failing with
```
  File "/home/datapastry/.cache/pypoetry/virtualenvs/rankeval-YsU-ilYr-py3.10/lib/python3.10/site-packages/guild/external/pip/_vendor/html5lib/_trie/_base.py", line 3, in <module>
    from collections import Mapping
ImportError: cannot import name 'Mapping' from 'collections' (/usr/lib/python3.10/collections/__init__.py)
```

I'm guessing that's covered by this issue. Do we have a workaround or a working branch we can use the in meantime?

**Comment by msarahan** [comment]
You can try the master branch with https://github.com/guildai/guildai/pull/324

the tests still fail with that pr but the normal code paths might work

**Comment by msarahan** [comment]
fixed in #341; closing
